src/validation.rs

```diff
@@ -256,55 +257,82 @@ impl Validator {
-                                    for (i, e) in expressions.iter().enumerate() {
-                                        // safe index of passed parameter
-                                        if let Ok((idx, _)) =
-                                            get_implicit_call_parameter(e, &declared_parameters, i)
-                                        {
-                                            passed_params_idx.push(idx);
-                                        }
-                                    }
-                                }
-                                _ => {
-                                    // safe index of passed parameter
-                                    if let Ok((idx, _)) =
-                                        get_implicit_call_parameter(s, &declared_parameters, 0)
-                                    {
-                                        passed_params_idx.push(idx);
-                                    }
-                                }
-                            }
-                        }
-                        // check if all inouts were passed to the pou call
-                        inouts.into_iter().for_each(|p| {
-                            if !passed_params_idx.contains(&(p.get_location_in_parent() as usize)) {
-                                self.stmt_validator.diagnostics.push(
-                                    Diagnostic::missing_inout_parameter(
-                                        p.get_name(),
-                                        operator.get_location(),
-                                    ),
+                if let Some(pou) = context.find_pou(operator) {
+                    let declared_parameters = context.index.get_declared_parameters(pou.get_name());
+                    let passed_parameters = parameters
+                        .as_ref()
+                        .as_ref()
+                        .map(ast::flatten_expression_list)
+                        .unwrap_or_default();
+
+                    let mut passed_params_idx = Vec::new();
+                    let mut are_implicit_parameters = true;
+                    // validate parameters
+                    for (i, p) in passed_parameters.iter().enumerate() {
+                        if let Ok((index, right, is_implicit)) =
+                            get_implicit_call_parameter(p, &declared_parameters, i)
+                        {
+                            // safe index of passed parameter
+                            passed_params_idx.push(index);
+                            let left_type = declared_parameters.get(index).map(|param| {
+                                context
+                                    .index
+                                    .get_effective_type_or_void_by_name(param.get_type_name())
+                            });
+                            let right_type = context.ast_annotation.get_type(right, context.index);
```

Comment:
I think we need to check against  `context.ast_annotation.get_type_hint(right, context.index).or_else(|| context.ast_annotation.get_type(...));` because the annotator may already requested a cast which will be reflected by the type-hint.

--> and we should get a helper in the annotation-struct since we do this again and again.

Reply:
we will always annotate if the cast can be done or not
-> trying to pass a string to int parameter -> we will have an annotation to the int type therefore we will always have the same type and never get the correct validation